elastic/kibana
# [Reporting] Add link to Kibana app from Reporting management UI + Design update

## Summary

Close https://github.com/elastic/kibana/issues/108498
Close https://github.com/elastic/kibana/issues/39620

This PR enhances the reporting management UI with the ability to link to different places in Kibana based on where a report was requested from. This will enable users to "debug" the places in Kibana that reporting will navigate to in their own browser since we no longer report the full URL in the server logs. Additionally, this grants the ability to view reporting as "snapshots" of different parts of Kibana UI. Users can navigate back to those states, update them and request a new report after they have navigated away from the UI.

## How to test

1. Start Kibana + ES on basic
2. Load some sample data into the cluster to get index patterns and at least one dashboard
3. Navigate to dashboard and generate a dashboard report
4. Go to reporting (check that there is a report in progress)
5. Attempt to download the report by clicking on the link once it is done
6. Navigate back to the Dashboard where you generated the report by clicking on the new action that links back to a Kibana app

## Additional changes

* Updated the Reporting management UI to be more in line with other similar management UIs
    * Refactored columns in table
    * Added a more mobile-friendly table
    * Refactored report names to be download links
    * Refactored actions in "Actions" column to be either link to place in Kibana or pre-existing information view
    * Action icons now only show the "i" icon, we use the "alert" icon to signal deprecations only
* Removed use of injected `intl` and rather use `@kbn/i18n` on various components
* Slight restructure of UI components folder - added new `components` folder to give a better indication of hierarchy of components
* Removed an instance of a user of `Boom` (related https://github.com/elastic/kibana/issues/105740)
* Moved the reporting redirect app to it's own top-level application. This means we have more control over what elements exactly will be rendered when navigating to the app so we can optimize and ensure a good UX when linking to a Kibana app.
* Removed a big-ass Jest snapshot in favour of a test that is more explicitly checking that jobs are present and ready to be downloaded

## Screenshots

### Before/current for reference

<img width="1643" alt="Screenshot 2021-09-07 at 17 19 10" src="https://user-images.githubusercontent.com/8155004/132370023-595897cc-6637-4f3f-be08-86138a48d06b.png">

### New table

<img width="1640" alt="Screenshot 2021-10-19 at 18 37 58" src="https://user-images.githubusercontent.com/8155004/137954379-84f5c6eb-7b08-41f6-9f08-b9e1ad546e72.png">

<img width="410" alt="Screenshot 2021-09-15 at 11 58 01" src="https://user-images.githubusercontent.com/8155004/133429881-2dacffd1-960d-46bf-9d11-b405da8458e4.png">

<img width="1082" alt="Screenshot 2021-09-15 at 12 14 30" src="https://user-images.githubusercontent.com/8155004/133429888-2732e2cd-754d-47d7-92cb-3c2e86e130c0.png">


### Mobile-friendly view
<img width="596" alt="Screenshot 2021-09-07 at 17 43 29" src="https://user-images.githubusercontent.com/8155004/132373726-96fcb23f-d149-4e37-8ca5-c163bdc8c94e.png">

### Interstitial redirect page for reporting (now that users will possibly see it)

#### If an error occurs while navigating

<img width="1031" alt="Screenshot 2021-10-19 at 18 32 12" src="https://user-images.githubusercontent.com/8155004/137954622-95bde9f6-71df-43c6-bef3-a8dfd56bb948.png">

## Release note

We updated the reporting management UI to place emphasis on report status and downloads. We also added a new capability to the report management UI: navigating back to the place a report was generated with the full state of the application at the time the report was requested. This is similar to the behaviour in the "Saved Objects" management UI.

<img width="1454" alt="Screenshot 2021-11-11 at 17 21 11" src="https://user-images.githubusercontent.com/8155004/141332425-ca81e911-4f11-4a07-b570-8cfcef8371ce.png">

### Checklist

    - Added a functional test for the new "navigate to Kibana app behaviour"
    - Added a Jest test to ensure that clicking the link is space aware
    
## Future work

It would be really nice to refactor the test file `x-pack/plugins/reporting/public/management/report_listing.test.tsx` since it has grown a bit large and a lot of the logic is setup of test data and the test harness.
